---
title: "Class 14:Transcriptomics and the Analysis of RNA-Seq Data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Import countData and colData
The data for this hands-on session comes from a published RNA-seq experiment where airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).

Glucocorticoids are used, for example, by people with asthma to reduce inflammation of the airways. The anti-inflammatory effects on airway smooth muscle (ASM) cells has been known for some time but the underlying molecular mechanisms are unclear.

```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE )

metadata <- read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```

Examining these files:

```{r}
head(counts)
```

```{r}
head(metadata)
```

```{r}
# Viewing entire object
View(counts)
View(metadata)
```

## Toy differential gene expression

Examine metadata to see the control samples. Finds the sample id for those labeled control. Then calculates the mean counts per gene across these samples. 

We want to calculate the average counts per gene for control samples and then for the drug treated samples.We will then compare these values to see if therei s a difference overallbetween control and treated samples for a given gene

```{r}
control <- metadata[metadata[,"dex"] == "control",]

control.mean <- rowSums(counts[,control$id]) / nrow(control)

names(control.mean) <- counts$ensgene

# Try to optimize this function
# metadata$id[metadata$dex == "control"]
# metadata[metadata$dex == "control",]$id

# Repeat the same process for the treated samples
treated <- metadata[metadata$dex == "treated",]

treated.mean <- rowSums(counts[,treated$id]) / nrow(treated)

names(treated.mean) <- counts$ensgene

# For bookkeeping purposes
meancounts <- data.frame(control.mean, treated.mean)
meancounts
```

Q1. How would you make the above code more robust? What would happen if you were to add more samples. Would the values obtained with the excat code above be correct?

A: Take out the `/4` and add `nrow(treated/control)`


Q2. Follow the same procedure for the treated samples (i.e. calculate the mean per gene accross drug treated samples and assign to a labeled vector called treated.mean)



# Show the sum of the mean counts across all genes for each group
```{r}
colSums(meancounts)
```

Q3. Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

```{r}
plot(meancounts, xlab = "Control", ylab = "Treated")
```

Plotting both axes on alog scale
```{r}
plot(meancounts, log = "xy", xlab = "log Control", ylab = "log Treated")
```

We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the `head()` or the `View()` function for example.

```{r}
meancounts$log2fc <- log2(meancounts[, "treated.mean"] / meancounts[, "control.mean"])

head(meancounts)
```

The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. Again inspect your result (and the intermediate steps) to see if things make sense to you

```{r}
zero.vals <- which(meancounts[, 1:2] == 0, arr.ind = TRUE)

to.rm <- unique(zero.vals[, 1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```



Remove any gene that has zero average counts
```{r}
x <- data.frame(c(1,1,10,0), c(0,1,10,0))
x
x == 0
ind <- unique(which(x == 0, arr.ind = TRUE)[,"row"])
x[-ind,]
```


Q4. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

A. arr.ind is used to provide the row/column where the function returns TRUE


A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)

sum(up.ind)
sum(down.ind)
```


Q5. Using the up.ind and down.ind vectors above can you determine how many up and down regulated genes we have at the greater than 2 fc level?

A. 250 and 367 respectively

## Do this properly with DESeq2

```{r}
library(DESeq2)
citation("DESeq2")
```


We will use the DESeqDataSetFromMatrix() function to build the required DESeqDataSet object and call it dds, short for our DESeqDataSet.

Setup input forDESeq by specificing countdata, metadat, and the experiment design (i.e. which comparisons we are interested in)

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design =~ dex, 
                              tidy = TRUE)

dds

dds <- DESeq(dds)
```

```{r}
res <- results(dds)
```

```{r}
plot(res$log2FoldChange, res$padj)
```

```{r}
plot(res$log2FoldChange,log(res$padj))
```

Let's plot this the way the rest of the world does!
```{r}
# Setup my point color vector
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange) > 2] <- "red"

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) >2)
mycols[inds] <- "blue"

# Volcano plot with customcolors

plot(res$log2FoldChange, -log(res$padj), col = mycols, ylab = "-Log(P-value)", xlab = "Log2(FoldChange)")

abline(v = c(-2,2), col = "gray", lty = 2)
abline(h = -log(0.1), col = "gray", lty = 2)
```


#DESeq pipeline

Next, let’s run the DESeq pipeline on the dataset, and reassign the resulting object back to the same variable. Before we start, dds is a bare-bones DESeqDataSet. The DESeq() function takes a DESeqDataSet and returns a DESeqDataSet, but with lots of other information filled in (normalization, dispersion estimates, differential expression results, etc). Notice how if we try to access these objects before running the analysis, nothing exists.

```{r}
sizeFactors(dds)
```

```{r}
dispersions(dds)
```

res


































